[core, lua, sql] Self-centered AoEs rework. Introduce mobskill finalizer.

sruon opened the pull request:
<!-- Remove space and place 'x' mark between square [] brackets or click the checkbox after saving to affirm the following points: -->
<!-- (it should look like this: - [x] I have ...) -->
**_I affirm:_**
- [x] I understand that if I do not agree to the following points by completing the checkboxes my PR will be ignored.
- [x] I understand I should leave resolving conversations to the LandSandBoat team so that reviewers won't miss what was said.
- [x] I have read and understood the [Contributing Guide](https://github.com/LandSandBoat/server/blob/base/CONTRIBUTING.md) and the [Code of Conduct](https://github.com/LandSandBoat/server/blob/base/CODE_OF_CONDUCT.md).
- [x] I have _**tested my code and the things my code has changed**_ since the last commit in the PR and will test after any later commits.

## What does this pull request do?
Part 1 of several to clean all of the mobskills jank.

- Undo several changes introduced in #5745 
  - Revert the changed skills to just ENEMY in their validTargets
    - Specifically the semantics of having SELF + ENEMY in validTargets hardly make sense and make the code hard to reason about. 
      - Using Touchdown as an example, it is a self-centered AoE with a certain radius that may or may not hit anyone. Tiamat is never targeting itself to deal damage, the only reason it was done was to ensure the animationsub update would occur in onMobWeaponskill which can be solved with finalizers instead.
  - Remove several if blocks that no longer make sense
- Introduce onMobSkillFinalize - a finalizer function that ~~always runs~~ runs when the skill has not been interrupted, independently of targets being found
  - This can be used to change an animationSub, kill mob on suicide etc even if the associated skill didn't hit anyone
  - I have at least 3 separate captures showing that certain mobskills do not need to hit anyone for their side effect to occur
- Introduce a skill flag SKILLFLAG_ALWAYS_ANIMATE to replace what used to be a SELF + ANY_ALLEGIANCE check for the sake of checking if the animation needed to complete regardless of the mobskill outcome
- Removes a range check from Tiamat for Touchdown as it is no longer necessary 

<!-- Describe what your PR does here. If it closes an existing issue, you can mention: "Closes #1234" and GitHub will link this PR to that issue. -->

## Steps to test these changes
Absolutely untested beyond Tiamat for now, just wanted to collect feedback for the time being. I still need to grab a few captures on retail first.

<!-- Clear and detailed steps to test your changes here -->

zach2good reviewed (approved):
A couple of things can be marked as const, but this is a good change 👍

sruon committed:
[core] Add SKILLFLAG_ALWAYS_ANIMATE and onMobSkillFinalize

- Add SKILLFLAG_ALWAYS_ANIMATE for skills that should complete
  animation even with no targets in range
- Add onMobSkillFinalize Lua callback that runs on mobskill state cleanup
  (if not interrupted)
- Fix self-centered AoE target validation and finding
- Update touchdown and suicide skills to use new callback for state changes
- Add warning when loading mobskills with both TARGET_SELF and TARGET_ENEMY
